### PR TITLE
RavenDB-7533 changes after latest PR review

### PR DESCRIPTION
**Pull request by myarichuk**

### CODE REVIEW

**Comment by myarichuk** [comment]
I've merged multiple commits into one for clarity

File: src/Raven.Client/Http/ClusterRequestExecutor.cs [hunk 1]
```diff
@@ -127,7 +125,7 @@ protected override Task PerformHealthCheck(ServerNode serverNode, JsonOperationC
             return true;
         }
 
-        protected override Task UpdateClientConfigurationAsync()
+        protected override Task UpdateClientConfigurationAsync(int? sessionId)
```
Comment (ayende) [inline src/Raven.Client/Http/ClusterRequestExecutor.cs:128]: Why do we need a session id here?

File: src/Raven.Client/Http/NodeSelector.cs [hunk 2]
```diff
@@ -50,18 +39,25 @@ public bool OnUpdateTopology(Topology topology, bool forceUpdate = false)
...
         }
 
-        public ServerNode GetCurrentNode()
+        public (int, ServerNode) GetCurrentNode()
         {
             if (Topology.Nodes.Count == 0)
                 ThrowEmptyTopology();
-            return Topology.Nodes[_currentNodeIndex];
+
+            return (Volatile.Read(ref _currentNodeIndex),Topology.Nodes[_currentNodeIndex]);
```
Comment (ayende) [inline src/Raven.Client/Http/NodeSelector.cs:54]: Here you are doing two _separate_ reads of the value.

File: src/Raven.Client/Http/NodeSelector.cs [hunk 3]
```diff
@@ -79,12 +75,20 @@ public void RestoreNodeIndex(int nodeIndex)
         public event Action<int> NodeSwitch;
 
         private readonly object _nodeIncrementSyncObj = new object();
-        protected void AtomicAdvanceNodeIndex()
+        public void AdvanceToNextNode(int? sessionId)
         {
             lock (_nodeIncrementSyncObj)
```
Comment (ayende) [inline src/Raven.Client/Http/NodeSelector.cs:80]: Why are you locking?!

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 4]
```diff
@@ -62,12 +62,12 @@ public class RequestExecutor : IDisposable
...
         private Timer _updateTopologyTimer;
 
-        protected INodeSelector _nodeSelector;
+        protected NodeSelector _nodeSelector;
 
         private TimeSpan? _defaultTimeout;
 
         //note: the condition for non empty nodes is precaution, should never happen..
-        public string Url => _nodeSelector?.GetCurrentNode()?.Url;
+        public string Url => _nodeSelector?.GetCurrentNode().Item2?.Url;
```
Comment (ayende) [inline src/Raven.Client/Http/RequestExecutor.cs:70]: Use named value here.

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 5]
```diff
@@ -171,22 +171,7 @@ public static RequestExecutor CreateForSingleNodeWithoutConfigurationUpdates(str
...
-                    return new RoundRobinNodeSelector(initialTopology);
-                case ReadBalanceBehavior.FastestNode:
-                //TODO : add here init of SLA load balancing node selector when it is implemented
-                default:
-                    throw new ArgumentOutOfRangeException(nameof(_readBalanceBehavior), _readBalanceBehavior, null);
-            }
-        }
-
-        protected virtual async Task UpdateClientConfigurationAsync()
+        protected virtual async Task UpdateClientConfigurationAsync(int? sessionId)
```
Comment (ayende) [inline src/Raven.Client/Http/RequestExecutor.cs:174]: Why do you need sessionId here?

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 6]
```diff
@@ -223,7 +209,7 @@ protected INodeSelector GetNodeSelector(Topology initialTopology)
             }
         }
 
-        public virtual async Task<bool> UpdateTopologyAsync(ServerNode node, int timeout)
+        public virtual async Task<bool> UpdateTopologyAsync(ServerNode node, int? nodeIndex, int timeout)
```
Comment (ayende) [inline src/Raven.Client/Http/RequestExecutor.cs:212]: Why do you need nodeIndex here?

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 7]
```diff
@@ -276,22 +264,55 @@ protected void DisposeAllFailedNodesTimers()
...
         }
 
-        public Task ExecuteAsync<TResult>(RavenCommand<TResult> command, JsonOperationContext context, CancellationToken token = default(CancellationToken))
+        public Task ExecuteAsync<TResult>(
+            RavenCommand<TResult> command, 
+            JsonOperationContext context, 
+            CancellationToken token = default(CancellationToken),
+            int? sessionId = null)
         {
+            if (command.GetType().Name.Contains("Query"))
```
Comment (ayende) [inline src/Raven.Client/Http/RequestExecutor.cs:281]: Why do we have this code here?

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 8]
```diff
@@ -627,11 +652,16 @@ private HttpRequestMessage CreateRequest<TResult>(ServerNode node, RavenCommand<
...
 
-        private async Task<bool> HandleUnsuccessfulResponse<TResult>(ServerNode chosenNode, int nodeIndex, JsonOperationContext context, RavenCommand<TResult> command, HttpRequestMessage request, HttpResponseMessage response, string url)
+        private async Task<bool> HandleUnsuccessfulResponse<TResult>(ServerNode chosenNode, int? nodeIndex, JsonOperationContext context, RavenCommand<TResult> command, HttpRequestMessage request, HttpResponseMessage response, string url, int? sessionId)
         {
             switch (response.StatusCode)
             {
                 case HttpStatusCode.NotFound:
+                    if (nodeIndex.HasValue)
+                    {
+                        _nodeSelector?.OnFailedRequest(nodeIndex.Value, sessionId);
```
Comment (ayende) [inline src/Raven.Client/Http/RequestExecutor.cs:662]: Why do we pass the session id here?

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 9]
```diff
@@ -627,11 +652,16 @@ private HttpRequestMessage CreateRequest<TResult>(ServerNode node, RavenCommand<
 
         public event Action<StringBuilder> AdditionalErrorInformation;
 
-        private async Task<bool> HandleUnsuccessfulResponse<TResult>(ServerNode chosenNode, int nodeIndex, JsonOperationContext context, RavenCommand<TResult> command, HttpRequestMessage request, HttpResponseMessage response, string url)
+        private async Task<bool> HandleUnsuccessfulResponse<TResult>(ServerNode chosenNode, int? nodeIndex, JsonOperationContext context, RavenCommand<TResult> command, HttpRequestMessage request, HttpResponseMessage response, string url, int? sessionId)
         {
             switch (response.StatusCode)
             {
                 case HttpStatusCode.NotFound:
+                    if (nodeIndex.HasValue)
```
Comment (ayende) [inline src/Raven.Client/Http/RequestExecutor.cs:660]: How is the nodeIndex relates to sessionId?

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 10]
```diff
@@ -641,11 +671,24 @@ private HttpRequestMessage CreateRequest<TResult>(ServerNode node, RavenCommand<
                     return true;
                 case HttpStatusCode.Forbidden:
                     throw new AuthorizationException("Forbidden access to " + chosenNode.Database + "@" + chosenNode.Url +", " + (Certificate == null ? "a certificate is required": Certificate.FriendlyName + " does not have permission to access it or is unknown"));
+                case HttpStatusCode.InternalServerError:
+                    var exceptionString = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+                    //try again, maybe it's transient error?
```
Comment (ayende) [inline src/Raven.Client/Http/RequestExecutor.cs:676]: Absolutely NOT.
You don't retry such things ever.

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 11]
```diff
@@ -926,14 +978,14 @@ public void Dispose()
             }
         }
 
-        public async Task<ServerNode> GetCurrentNode()
+        public async Task<(int, ServerNode)> GetCurrentNode()
```
Comment (ayende) [inline src/Raven.Client/Http/RequestExecutor.cs:981]: Use named tuples.

**Review by ayende — CHANGES_REQUESTED** [review]

**Comment by myarichuk** [comment]
I will reopen with another PR

**Pull request closed without merging**